### PR TITLE
Split AstToExecutor from dbgFuncCoprocessor.

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609 

Problem Summary:
Need to provide Ast to tipb::Executor helper functions for Interpreter test framework, but all of them are in .cpp file.

### What is changed and how it works?
Move  Ast to tipb::Executor helper functions to astToExecutor.h
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Debug/AstToExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,1516 @@
...
+        expr->set_tp(tipb::ColumnRef);
+        *(expr->mutable_field_type()) = columnInfoToFieldType((*ft).second);
+        expr->mutable_field_type()->set_collate(collator_id);
+        WriteBufferFromOwnString ss;
+        encodeDAGInt64(ft - input.begin(), ss);
+        expr->set_val(ss.releaseStr());
+    }
+    else if (ASTFunction * func = typeid_cast<ASTFunction *>(ast.get()))
+    {
+        /// aggregation function is handled in Aggregation, so just treated as a column
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.cpp:279]: how about move this branch into another function?

File: dbms/src/Debug/AstToExecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,1516 @@
...
+            else
+                return field_name.first == column_name.first && field_name.second == column_name.second;
+        });
+        if (ft == input.end())
+            throw Exception("No such column " + id->getColumnName(), ErrorCodes::NO_SUCH_COLUMN_IN_TABLE);
+        ci = ft->second;
+    }
+    else if (ASTFunction * func = typeid_cast<ASTFunction *>(ast.get()))
+    {
+        /// check function
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.cpp:563]: ditto

File: dbms/src/Debug/AstToExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,1516 @@
...
+            }
+            return;
+        }
+    }
+    conditions.push_back(ast);
+    compileExpr(input, ast);
+}
+} // namespace
+
+String LOCAL_HOST = "127.0.0.1:3930";
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.cpp:736]: move to anonymous namespace?

File: dbms/src/Debug/AstToExecutor.h [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    Executor(size_t & index_, String && name_, const DAGSchema & output_schema_)
+        : index(index_)
+        , name(std::move(name_))
+        , output_schema(output_schema_)
+    {
+        index_++;
+    }
+    virtual bool toTiPBExecutor(tipb::Executor * tipb_executor, uint32_t collator_id, const MPPInfo & mpp_info, const Context & context)
+        = 0;
+    virtual void toMPPSubPlan(size_t & executor_index, const DAGProperties & properties, std::unordered_map<String, std::pair<std::shared_ptr<ExchangeReceiver>, std::shared_ptr<ExchangeSender>>> & exchange_map)
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.h:110]: is `executor_index` expected to be mutable to derived classes of `Executor`?

File: dbms/src/Debug/AstToExecutor.h [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    }
+    virtual ~Executor() = default;
+};
+
+struct ExchangeSender : Executor
+{
+    tipb::ExchangeType type;
+    TaskMetas task_metas;
+    std::vector<size_t> partition_keys;
+    ExchangeSender(size_t & index, const DAGSchema & output, tipb::ExchangeType type_, const std::vector<size_t> & partition_keys_ = {})
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.h:122]: is `index` mutable?

File: dbms/src/Debug/AstToExecutor.h [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        , partition_keys(partition_keys_)
+    {}
+    void columnPrune(std::unordered_set<String> &) override { throw Exception("Should not reach here"); }
+    bool toTiPBExecutor(tipb::Executor * tipb_executor, uint32_t collator_id, const MPPInfo & mpp_info, const Context & context) override;
+};
+
+struct ExchangeReceiver : Executor
+{
+    TaskMetas task_metas;
+    ExchangeReceiver(size_t & index, const DAGSchema & output)
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.h:134]: ditto

File: dbms/src/Debug/AstToExecutor.h [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    {}
+    void columnPrune(std::unordered_set<String> &) override { throw Exception("Should not reach here"); }
+    bool toTiPBExecutor(tipb::Executor * tipb_executor, uint32_t collator_id, const MPPInfo & mpp_info, const Context &) override;
+};
+
+struct TableScan : public Executor
+{
+    TableInfo table_info;
+    /// used by column pruner
+    TableScan(size_t & index_, const DAGSchema & output_schema_, TableInfo & table_info_)
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.h:145]: ditto, the same as `table_info_`.

File: dbms/src/Debug/AstToExecutor.h [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    void columnPrune(std::unordered_set<String> & used_columns) override;
+    bool toTiPBExecutor(tipb::Executor * tipb_executor, uint32_t, const MPPInfo &, const Context &) override;
+    void toMPPSubPlan(size_t &, const DAGProperties &, std::unordered_map<String, std::pair<std::shared_ptr<ExchangeReceiver>, std::shared_ptr<ExchangeSender>>> &) override
+    {}
+};
+
+struct Selection : public Executor
+{
+    std::vector<ASTPtr> conditions;
+    Selection(size_t & index_, const DAGSchema & output_schema_, std::vector<ASTPtr> && conditions_)
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.h:158]:
```suggestion
    Selection(size_t & index_, const DAGSchema & output_schema_, std::vector<ASTPtr> conditions_)
```

File: dbms/src/Debug/AstToExecutor.h [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    {}
+    bool toTiPBExecutor(tipb::Executor * tipb_executor, uint32_t collator_id, const MPPInfo & mpp_info, const Context & context) override;
+    void columnPrune(std::unordered_set<String> & used_columns) override;
+};
+
+struct TopN : public Executor
+{
+    std::vector<ASTPtr> order_columns;
+    size_t limit;
+    TopN(size_t & index_, const DAGSchema & output_schema_, std::vector<ASTPtr> && order_columns_, size_t limit_)
```
Comment (fuzhe1989) [inline dbms/src/Debug/AstToExecutor.h:170]:
```suggestion
    TopN(size_t & index_, const DAGSchema & output_schema_, std::vector<ASTPtr> order_columns_, size_t limit_)
```

File: dbms/src/Debug/AstToExecutor.cpp (hunk 1 above)
Comment (ywqzzy) [reply dbms/src/Debug/AstToExecutor.cpp:279]: > how about move this branch into another function?

Good idea. I will refactor them all.

**Review by SeaRise — APPROVED** [review]

File: dbms/src/Debug/astToExecutor.cpp [hunk 10]
```diff
@@ -0,0 +1,1519 @@
...
+
+namespace Debug
+{
+void setServiceAddr(const std::string & addr)
+{
+    LOCAL_HOST = addr;
+}
+} // namespace Debug
+
+std::pair<String, String> splitQualifiedName(String s)
```
Comment (yibin87) [inline dbms/src/Debug/astToExecutor.cpp:752]: Change parameter to const String & s ?

File: dbms/src/Debug/AstToExecutor.h (hunk 5 above)
Comment (ywqzzy) [reply dbms/src/Debug/AstToExecutor.h]: it is mutable

File: dbms/src/Debug/astToExecutor.cpp (hunk 10 above)
Comment (ywqzzy) [reply dbms/src/Debug/astToExecutor.cpp]: done.

File: dbms/src/Debug/AstToExecutor.h (hunk 7 above)
Comment (ywqzzy) [reply dbms/src/Debug/AstToExecutor.h]: done

File: dbms/src/Debug/AstToExecutor.h (hunk 8 above)
Comment (ywqzzy) [reply dbms/src/Debug/AstToExecutor.h]: done.

File: dbms/src/Debug/AstToExecutor.h (hunk 9 above)
Comment (ywqzzy) [reply dbms/src/Debug/AstToExecutor.h]: done

File: dbms/src/Debug/AstToExecutor.cpp (hunk 2 above)
Comment (ywqzzy) [reply dbms/src/Debug/AstToExecutor.cpp]: done

File: dbms/src/Debug/AstToExecutor.cpp (hunk 3 above)
Comment (ywqzzy) [reply dbms/src/Debug/AstToExecutor.cpp]: done. I have moved it into namespace db::debug.

**Review by yibin87 — APPROVED** [review]
LGTM

**Review by fuzhe1989 — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
Others LGTM

**Pull request merged**